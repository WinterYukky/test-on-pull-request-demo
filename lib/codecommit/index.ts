export * from "aws-cdk-lib/aws-codecommit";
export * from "./approval-rule-template";
