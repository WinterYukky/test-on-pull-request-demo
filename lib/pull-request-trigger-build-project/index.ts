export * from "./pull-request-trigger-build-project";
