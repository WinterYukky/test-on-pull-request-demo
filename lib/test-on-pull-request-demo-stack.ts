import * as cdk from "aws-cdk-lib";
import { Asset } from "aws-cdk-lib/aws-s3-assets";
import { Construct } from "constructs";
import * as codecommit from "./codecommit";
import { PullRequestTriggerBuildProject } from "./pull-request-trigger-build-project";

export class TestOnPullRequestDemoStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const gitRepository = new codecommit.Repository(this, "Repository", {
      repositoryName: "PullReqeustApproveDemo",
      // Initialize Git repository with this AWS CDK project code for testing.
      code: codecommit.Code.fromAsset(
        new Asset(this, "CodeAsset", {
          path: ".",
          exclude: [".git", "cdk.out", "node_modules"],
        })
      ),
    });
    new PullRequestTriggerBuildProject(this, "PullRequestTriggerBuildProject", {
      repository: gitRepository,
      branchFilters: ["refs/heads/main"],
      prebuildCommands: ["node --version", "npm ci"],
      commands: ["npm run test"],
    });
  }
}
