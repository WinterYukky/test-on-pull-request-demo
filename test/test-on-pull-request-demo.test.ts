import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { TestOnPullRequestDemoStack } from "../lib/test-on-pull-request-demo-stack";

test("snapshot test", () => {
  const app = new App();
  const stack = new TestOnPullRequestDemoStack(app, "MyTestStack");
  const template = Template.fromStack(stack).toJSON();
  expect(template).toMatchSnapshot();
});
