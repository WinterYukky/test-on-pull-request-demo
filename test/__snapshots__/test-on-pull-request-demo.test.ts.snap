// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PullRequestTriggerBuildProjectApprovalRuleTemplate1F686289": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "PullRequestTriggerBuildProjectApprovalRuleTemplateCustomResourcePolicy837FF73C",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"@aws-sdk/client-codecommit","action":"CreateApprovalRuleTemplateCommand","parameters":{"approvalRuleTemplateName":"MyTestStackPullRequestTriggerBuildProjectApprovalRuleTemplate0CF9F3C0","approvalRuleTemplateContent":"{\\"Version\\":\\"2018-11-08\\",\\"DestinationReferences\\":[\\"refs/heads/main\\"],\\"Statements\\":[{\\"Type\\":\\"Approvers\\",\\"ApprovalPoolMembers\\":[\\"CodeCommitApprovers:",
              {
                "Ref": "PullRequestTriggerBuildProjectProjctRole1AA37B71",
              },
              "/*\\"],\\"NumberOfApprovalsNeeded\\":1}]}"},"physicalResourceId":{"responsePath":"approvalRuleTemplate.approvalRuleTemplateId"}}",
            ],
          ],
        },
        "Delete": "{"service":"@aws-sdk/client-codecommit","action":"DeleteApprovalRuleTemplateCommand","parameters":{"approvalRuleTemplateName":"MyTestStackPullRequestTriggerBuildProjectApprovalRuleTemplate0CF9F3C0"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": {
          "Fn::Join": [
            "",
            [
              "{"service":"@aws-sdk/client-codecommit","action":"UpdateApprovalRuleTemplateContentCommand","parameters":{"approvalRuleTemplateName":"MyTestStackPullRequestTriggerBuildProjectApprovalRuleTemplate0CF9F3C0","newRuleContent":"{\\"Version\\":\\"2018-11-08\\",\\"DestinationReferences\\":[\\"refs/heads/main\\"],\\"Statements\\":[{\\"Type\\":\\"Approvers\\",\\"ApprovalPoolMembers\\":[\\"CodeCommitApprovers:",
              {
                "Ref": "PullRequestTriggerBuildProjectProjctRole1AA37B71",
              },
              "/*\\"],\\"NumberOfApprovalsNeeded\\":1}]}"},"physicalResourceId":{"responsePath":"approvalRuleTemplate.approvalRuleTemplateId"}}",
            ],
          ],
        },
      },
      "Type": "Custom::ApprovalRuleTemplate",
      "UpdateReplacePolicy": "Delete",
    },
    "PullRequestTriggerBuildProjectApprovalRuleTemplateAssociate052FDBDD7": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "PullRequestTriggerBuildProjectApprovalRuleTemplateAssociate0CustomResourcePolicy22DCCB18",
        "PullRequestTriggerBuildProjectApprovalRuleTemplateCustomResourcePolicy837FF73C",
        "PullRequestTriggerBuildProjectApprovalRuleTemplate1F686289",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"@aws-sdk/client-codecommit","action":"AssociateApprovalRuleTemplateWithRepositoryCommand","parameters":{"approvalRuleTemplateName":"MyTestStackPullRequestTriggerBuildProjectApprovalRuleTemplate0CF9F3C0","repositoryName":"",
              {
                "Fn::GetAtt": [
                  "Repository22E53BBD",
                  "Name",
                ],
              },
              ""},"physicalResourceId":{"id":"Associate0"}}",
            ],
          ],
        },
        "Delete": {
          "Fn::Join": [
            "",
            [
              "{"service":"@aws-sdk/client-codecommit","action":"DisassociateApprovalRuleTemplateFromRepositoryCommand","parameters":{"approvalRuleTemplateName":"MyTestStackPullRequestTriggerBuildProjectApprovalRuleTemplate0CF9F3C0","repositoryName":"",
              {
                "Fn::GetAtt": [
                  "Repository22E53BBD",
                  "Name",
                ],
              },
              ""}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": {
          "Fn::Join": [
            "",
            [
              "{"service":"@aws-sdk/client-codecommit","action":"AssociateApprovalRuleTemplateWithRepositoryCommand","parameters":{"approvalRuleTemplateName":"MyTestStackPullRequestTriggerBuildProjectApprovalRuleTemplate0CF9F3C0","repositoryName":"",
              {
                "Fn::GetAtt": [
                  "Repository22E53BBD",
                  "Name",
                ],
              },
              ""},"physicalResourceId":{"id":"Associate0"}}",
            ],
          ],
        },
      },
      "Type": "Custom::AssociateApprovalRuleTemplateWithRepository",
      "UpdateReplacePolicy": "Delete",
    },
    "PullRequestTriggerBuildProjectApprovalRuleTemplateAssociate0CustomResourcePolicy22DCCB18": {
      "DependsOn": [
        "PullRequestTriggerBuildProjectApprovalRuleTemplateCustomResourcePolicy837FF73C",
        "PullRequestTriggerBuildProjectApprovalRuleTemplate1F686289",
      ],
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codecommit:AssociateApprovalRuleTemplateWithRepository",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "codecommit:DisassociateApprovalRuleTemplateFromRepository",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PullRequestTriggerBuildProjectApprovalRuleTemplateAssociate0CustomResourcePolicy22DCCB18",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PullRequestTriggerBuildProjectApprovalRuleTemplateCustomResourcePolicy837FF73C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codecommit:CreateApprovalRuleTemplate",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "codecommit:UpdateApprovalRuleTemplateContent",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "codecommit:DeleteApprovalRuleTemplate",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PullRequestTriggerBuildProjectApprovalRuleTemplateCustomResourcePolicy837FF73C",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PullRequestTriggerBuildProjectProjct8A555D34": {
      "Properties": {
        "Artifacts": {
          "Type": "NO_ARTIFACTS",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:7.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": {
          "Fn::GetAtt": [
            "PullRequestTriggerBuildProjectProjctRole1AA37B71",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "version: 0.2
phases:
  install:
    runtime-versions:
      nodejs: 20.x
  pre_build:
    commands:
      - node --version
      - npm ci
  build:
    commands:
      - npm run test
      - REVISION_ID=$(aws codecommit get-pull-request --pull-request-id $PULL_REQUEST_ID | jq -r '.pullRequest.revisionId')
      - aws codecommit update-pull-request-approval-state --pull-request-id $PULL_REQUEST_ID --revision-id $REVISION_ID --approval-state APPROVE --region $AWS_REGION
",
          "Location": {
            "Fn::GetAtt": [
              "Repository22E53BBD",
              "CloneUrlHttp",
            ],
          },
          "Type": "CODECOMMIT",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "PullRequestTriggerBuildProjectProjctEventsRoleC0C94D46": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PullRequestTriggerBuildProjectProjctEventsRoleDefaultPolicy67134DCA": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codebuild:StartBuild",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "PullRequestTriggerBuildProjectProjct8A555D34",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PullRequestTriggerBuildProjectProjctEventsRoleDefaultPolicy67134DCA",
        "Roles": [
          {
            "Ref": "PullRequestTriggerBuildProjectProjctEventsRoleC0C94D46",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PullRequestTriggerBuildProjectProjctRole1AA37B71": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PullRequestTriggerBuildProjectProjctRoleDefaultPolicy68CD0E93": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codecommit:GitPull",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "Repository22E53BBD",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "PullRequestTriggerBuildProjectProjct8A555D34",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "PullRequestTriggerBuildProjectProjct8A555D34",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    {
                      "Ref": "PullRequestTriggerBuildProjectProjct8A555D34",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "codecommit:CreatePullRequestApprovalRule",
                "codecommit:GetPullRequest",
                "codecommit:PostCommentForPullRequest",
                "codecommit:UpdatePullRequestApprovalState",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "Repository22E53BBD",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PullRequestTriggerBuildProjectProjctRoleDefaultPolicy68CD0E93",
        "Roles": [
          {
            "Ref": "PullRequestTriggerBuildProjectProjctRole1AA37B71",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PullRequestTriggerBuildProjectPullRequestTriggerRuleD92EC872": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "destinationReference": [
              "refs/heads/main",
            ],
            "event": [
              "pullRequestCreated",
              "pullRequestSourceBranchUpdated",
            ],
            "isMerged": [
              "False",
            ],
            "repositoryNames": [
              {
                "Fn::GetAtt": [
                  "Repository22E53BBD",
                  "Name",
                ],
              },
            ],
          },
          "source": [
            "aws.codecommit",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "PullRequestTriggerBuildProjectProjct8A555D34",
                "Arn",
              ],
            },
            "Id": "Target0",
            "InputTransformer": {
              "InputPathsMap": {
                "detail-destinationCommit": "$.detail.destinationCommit",
                "detail-pullRequestId": "$.detail.pullRequestId",
                "detail-sourceCommit": "$.detail.sourceCommit",
              },
              "InputTemplate": "{
	"sourceVersion": "<detail-sourceCommit>",
	"environmentVariablesOverride": [
		{
			"name": "DESTINATION_COMMIT_ID",
			"type": "PLAINTEXT",
			"value": "<detail-destinationCommit>"
		},
		{
			"name": "SOURCE_COMMIT_ID",
			"type": "PLAINTEXT",
			"value": "<detail-sourceCommit>"
		},
		{
			"name": "PULL_REQUEST_ID",
			"type": "PLAINTEXT",
			"value": "<detail-pullRequestId>"
		}
	]
}",
            },
            "RoleArn": {
              "Fn::GetAtt": [
                "PullRequestTriggerBuildProjectProjctEventsRoleC0C94D46",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "Repository22E53BBD": {
      "Properties": {
        "Code": {
          "S3": {
            "Bucket": {
              "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
            },
            "Key": "HASH-REPLACED.zip",
          },
        },
        "RepositoryName": "PullReqeustApproveDemo",
      },
      "Type": "AWS::CodeCommit::Repository",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
